domain/application/state/types.go

```diff
@@ -411,6 +412,31 @@ type setCharmRelation struct {
 	ScopeID   int    `db:"scope_id"`
 }
 
+// relationInfo represents metadata and configuration details for an existing
+// relation within an application.
+type relationInfo struct {
+	CharmUUID string `db:"charm_uuid"`
+	Name      string `db:"name"`
+	Role      string `db:"role"`
+	Interface string `db:"interface"`
+	Optional  bool   `db:"optional"`
+	Capacity  int    `db:"capacity"`
+	Scope     string `db:"scope"`
+	Count     int    `db:"count"`
+}
+
+// Relation transforms the relationInfo data into an internalcharm.Relation structure.
+func (ri relationInfo) Relation() internalcharm.Relation {
+	return internalcharm.Relation{
+		Name:      ri.Name,
+		Role:      internalcharm.RelationRole(ri.Role),
+		Interface: ri.Interface,
+		Optional:  ri.Optional,
+		Limit:     ri.Capacity,
+		Scope:     internalcharm.RelationScope(ri.Scope),
+	}
+}
```

Comment:
We have been moving out of these mapping methods here in the types.go and move them to the state layer as encode* or decode*. I know that we still have plenty of negative examples, but at least we should go further with one pattern only.